### PR TITLE
feat: Migrate Table to fiori3

### PR DESCRIPTION
**Pull request by JKMarkowski**
## Related Issue
Related to https://github.com/SAP/fundamental-styles/issues/914

## Descriptions
There are included
- Base Table Component - Header, Body, Footer + sizes
- Interactive states(hover, active, selected)
- Table with checkboxes + sizes
- Highlight Row
- Table with pagination
- Responsive table + pop-in mode
- Filter / Sort of table
- Navigation indicator inside table
- Fixed Column

## Breaking Changes
Added Class:
Modifiers:
- `fd-fable` + `(__header/__body/__cell)` + `(--no-vertical-borders/--no-horizontal-borders)`
- `fd-table` + `(__row/_cell)` + `(--activable/--hoverable)`
- `fd-table__cell--checkbox`

- `fd-table__cell` + `(--success/--error/--warning--information)`
- `fd-table__cell--status-indicator`+ `(--success/--error/--warning--information)`

Responsive:
- `fd-table__row--main`
- `fd-table__row--secondary`

Removed Classes:
Elements:
- `fd-table__sort-column(--asc/--desc)`
- `fd-table__context-menu`
- `fd-table__context-menu-label`
Modifiers: 
- `fd-table__row(--success/--error/--warning/--information)`

### Before:
image
image

### After:
image

image

#### Please check whether the PR fulfills the following requirements

- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-styles/wiki/PR-Review-Checklist

### CODE REVIEW

**Comment by mikerodonnell89** [comment]
Too many cells in the footer on this example
Screen Shot 2020-05-21 at 1 24 21 PM

**Comment by mikerodonnell89** [comment]
Strange focus outline for condensed checkboxes
Screen Shot 2020-05-21 at 1 24 21 PM

**Comment by mikerodonnell89** [comment]
Table cell contents aligned left in RTL
Screen Shot 2020-05-21 at 1 38 31 PM

File: src/table.scss [hunk 1]
```diff
@@ -11,32 +11,29 @@
...
-  $fd-table-fixed-column-cell-color: fd-color("background", 2);
-
-  --fd-table-border-color: var(--fd-color-neutral-2);
-  --fd-table-background-color: var(--fd-color-background-2);
-  --fd-table-link-color: var(--fd-color-action-1);
-  --fd-table-header-color: var(--fd-color-text-3);
-  --fd-table-header-background-color: var(--fd-color-neutral-1);
+
+  $fd-table-cell-padding: 0.5rem;
+  $fd-table-header-height: 2.75rem;
```
Comment (mikerodonnell89) [inline src/table.scss:33]: Weird name for this variable because it is only used in cell selectors

File: src/table.scss [hunk 2]
```diff
@@ -48,258 +45,173 @@ $block: #{$fd-namespace}-table;
...
-  &__row {
-    transition: background-color $fd-table-transition-params;
+  &__header {
+    .#{$block}__row {
+      border: none;
 
-    @include fd-hover() {
-      @include fd-var-color("background-color", $fd-table-row-background-color--hover, --fd-color-background-hover);
+      &:hover {
+        background-color: inherit;
```
Comment (mikerodonnell89) [inline src/table.scss:67]: is this needed for something that I'm missing?

File: src/table.scss [hunk 3]
```diff
@@ -48,258 +45,173 @@ $block: #{$fd-namespace}-table;
...
-          @include fd-var-color("border-color", fd-color('status', 1), --fd-color-status-1);
-        }
-      }
+  &__cell {
+    text-align: left;
+    height: $fd-table-header-height;
+    padding: 0 $fd-table-cell-padding;
+    font-family: var(--sapFontFamily);
+    font-size: var(--sapFontSize);
+    font-weight: normal;
```
Comment (InnaAtanasova) [inline src/table.scss:90]: normal font-weight is coming with the fd-reset

File: src/table.scss [hunk 4]
```diff
@@ -48,258 +45,173 @@ $block: #{$fd-namespace}-table;
...
-  @include fd-var-color("border-color", $fd-table-border-color, --fd-table-border-color);
+  &__icon {
+    color: var(--sapContent_NonInteractiveIconColor);
+    font-size: var(--sapFontSmallSize);
+    vertical-align: middle;
+  }
 
-  &__row {
-    transition: background-color $fd-table-transition-params;
+  &__header {
```
Comment (InnaAtanasova) [inline src/table.scss:62]: I think your header is missing the border

File: src/table.scss [hunk 5]
```diff
@@ -48,258 +45,173 @@ $block: #{$fd-namespace}-table;
...
-      }
+  &__cell {
+    text-align: left;
+    height: $fd-table-header-height;
+    padding: 0 $fd-table-cell-padding;
+    font-family: var(--sapFontFamily);
+    font-size: var(--sapFontSize);
+    font-weight: normal;
+    color: inherit;
+    vertical-align: middle;
```
Comment (InnaAtanasova) [inline src/table.scss:92]: I don't think this is doing anything

**Comment by JKMarkowski** [comment]
Hi @InnaAtanasova 
> 1. I don't think the cells should have active state. They have hover, selected, hover+selected but didn't see the active (I might have missed it). The cells get focus tho (seen in the demo app)
- There is down state, which I understand as a active. You can take a look at activable cells/rows example on preview.

> 2. The RTL support is "missing" - the text should not be left-aligned and the first column padding (1rem) should go to the other side
- Fixed

> 3. Items that are navigatable have different color
>    Screen Shot 2020-05-21 at 5 18 46 PM
- It's shown only on some images, I don't know what to do with links actually links got own colors. Maybe I should have something like `fd-table__text` to keep color inherited.

> 1. Freeze - shouldn't we have a horizontal/vertical scroll when we freeze a certain row or column? For example, if we freeze the first column, we will be able to horizontally scroll in order to see all the columns (this is usually when we can't fit all the columns in the vw)

We have Fixed Column now. It's in last example, I would also add something like fixed row.

File: src/table.scss (hunk 1 above)
Comment (JKMarkowski) [reply src/table.scss]: Naming changed

**Comment by salarenko** [comment]
Nice work.
I'm not sure if this composition approach is correct in terms of interactivity and other features. Solid base should be a table with basic styles. Currently we bring with base table additional behaviors like:
- Marking row on hover
- Interactive table headers
I'd rather consider those as extensions of basic table.

**Comment by prsdthkr** [comment]
Great work :D
I noticed that the table here behaves like a [responsive table](https://ui5.sap.com/#/entity/sap.m.Table). If we were to follow the spec for the [grid table](https://ui5.sap.com/#/entity/sap.ui.table.Table), the columns will not be responsive by default as seen in [this example](https://ui5.sap.com/#/entity/sap.ui.table.Table/sample/sap.ui.table.sample.Basic). Instead, a scrollable view is provided (in smaller viewports or at 200% zoom). Also there are 3 modes for column width behavior as [seen here](https://ui5.sap.com/#/entity/sap.ui.table.Table/sample/sap.ui.table.sample.Resizing): static, flexible, and mixed.

I understand if this is out of scope for this PR.

**Comment by JKMarkowski** [comment]
Hi @prsdthkr,

Those features with responsive behaviour are up to application specification. If user adds some `min-width` to cells and adds wrapper with `overflow-x` it's possible to achieve. I added example with fixed column, which actually shows how it can be used.

Also there is responsive table example with `pop-in` mode, which changes view of table(it has modified markup a little bit)
image

I'm not sure what about the static, flexible, and mixed. It's not mentioned anywhere in fiori3 specs. We can add this example with dummy cells in future. Thanks for taking a look at it.

**Comment by InnaAtanasova** [comment]
Documentation:
The checkbox needs to be checked with selected state
Screen Shot 2020-05-28 at 3 12 46 PM

Responsive table:
- missing border 
Screen Shot 2020-05-28 at 3 11 56 PM

- prob not achievable with pure css but the whole item should get hover and selected state:
Screen Shot 2020-05-28 at 3 11 08 PM
Screen Shot 2020-05-28 at 3 10 46 PM

Navigation:
- I think the button should not have states. The states should be applied to the cell
Screen Shot 2020-05-28 at 3 08 19 PM

File: src/table.scss [hunk 6]
```diff
@@ -3,355 +3,415 @@
...
+  &__icon {
+    color: var(--sapContent_NonInteractiveIconColor);
+    font-size: var(--sapFontSmallSize);
 
-  border-style: solid;
-  border-width: $fd-table-border-width;
+    &--navigation {
+      font-size: var(--sapFontSmallSize);
+      width: 2rem;
+      min-width: 2rem;
```
Comment (InnaAtanasova) [inline src/table.scss:75]: should be put in a variable

File: src/table.scss [hunk 7]
```diff
@@ -3,355 +3,415 @@
...
+      color: var(--sapInformationColor);
+    }
 
-      .#{$block}__cell:first-child {
-        &::before {
-          @include fd-var-color("border-color", fd-color('status', 1), --fd-color-status-1);
-        }
-      }
+    &--checkbox {
+      width: 10px;
```
Comment (InnaAtanasova) [inline src/table.scss:133]: rem value

File: src/table.scss [hunk 8]
```diff
@@ -3,355 +3,415 @@
...
+      padding: 0;
+      margin: 0;
     }
 
-    &--warning {
-      @include fd-var-color("background-color", fd-color('background', 4), --fd-color-background-4);
+    &--fit-content {
+      width: 10px;
```
Comment (InnaAtanasova) [inline src/table.scss:139]: rem value

File: src/table.scss [hunk 9]
```diff
@@ -3,355 +3,415 @@
...
+    }
 
-      .#{$block}__cell:first-child {
-        &::before {
-          @include fd-var-color("border-color", fd-color('status', 2), --fd-color-status-2);
-        }
-      }
+    &--no-padding {
+      padding: 0;
+      margin: 0;
```
Comment (InnaAtanasova) [inline src/table.scss:146]: since we see this twice in the component you can create a local reset mixin and reuse it

File: src/table.scss [hunk 10]
```diff
@@ -3,355 +3,415 @@
...
-    @include fd-reset();
-    @include fd-type("-1");
-    @include fd-var-color("color", $fd-table-header-color, --fd-table-header-color);
-    @include fd-var-color("background-color", $fd-table-header-background-color, --fd-table-header-background-color);
+    &--fixed {
+      left: 0;
+      position: absolute;
+      background-color: $fd-table-basic-background;
+      display: flex;
+      align-items: center;
```
Comment (InnaAtanasova) [inline src/table.scss:175]: `fd-flex-vertical-center` mixin

**Comment by InnaAtanasova** [comment]
`--borderless-x` and `--borderless-y` instead of `--no-vertical-borders` and `--no-horizontal-borders` ? But we better discuss it with the rest of the team

**Comment by InnaAtanasova** [comment]
Screen Shot 2020-05-28 at 3 43 16 PM

The navigated indicator should be on the other side and the icon should point to the other direction

**Comment by InnaAtanasova** [comment]
Screen Shot 2020-05-28 at 3 42 48 PM

**Comment by InnaAtanasova** [comment]
Screen Shot 2020-05-28 at 3 41 51 PM

In RTL the fixed column should go on the right side

**Comment by InnaAtanasova** [comment]
Screen Shot 2020-05-28 at 3 40 33 PM

RTL doc example - arrows should be mirrored

**Comment by JKMarkowski** [comment]
@InnaAtanasova 

> Documentation:
> The checkbox needs to be checked with selected state
> Screen Shot 2020-05-28 at 3 12 46 PM

Done

> Responsive table:
> 
> * missing border
>   Screen Shot 2020-05-28 at 3 11 56 PM
It shouldn't have border (
image
image


> * prob not achievable with pure css but the whole item should get hover and selected state:
>   Screen Shot 2020-05-28 at 3 11 08 PM
>   Screen Shot 2020-05-28 at 3 10 46 PM
TODO

> 
> Navigation:
> 
> * I think the button should not have states. The states should be applied to the cell
>   Screen Shot 2020-05-28 at 3 08 19 PM
I don't think so, User shouldn't perform action on triggering cell, but button

**Comment by stefanoScalzo** [comment]
Screen Shot 2020-06-01 at 10 10 31 AM
The toolbar border is overlapping with the table border

**Comment by JKMarkowski** [comment]
@stefanoScalzo Should I be considered as something wrong?

**Comment by stefanoScalzo** [comment]
> @stefanoScalzo Should I be considered as issue?

True, something to discuss because I think we need to give our components the option to not include the border for cases like this? either that or we use a class that positions the elements lower with a smaller z index to hide it, but this may not work due to some using box-shadow. We need to discuss it.

**Comment by stefanoScalzo** [comment]
For: Responsive Table - Pop-in Mode with Checkboxes
When clicking on the checkbox the whole table gets highlighted but when clicking the table nothing gets checked.

**Comment by JKMarkowski** [comment]
> For: Responsive Table - Pop-in Mode with Checkboxes
> When clicking on the checkbox the whole table gets highlighted but when clicking the table nothing gets checked.

there is no way to make it work on pure css, I can't wrap row/cell to label.

**Comment by stefanoScalzo** [comment]
Can you include Navigated Item Indicator Bar in your responsive table example please

**Comment by stefanoScalzo** [comment]
Screen Shot 2020-06-01 at 11 09 51 AM
Overflow property not working properly

**Comment by mikerodonnell89** [comment]
Filter section has compact buttons but the rest of the table is cozy - intentional?

**Comment by JKMarkowski** [comment]
@mikerodonnell89 It's done by accident, I will change it

**Comment by InnaAtanasova** [comment]
> 
> > Responsive table:
> > 
> > * missing border
> >   Screen Shot 2020-05-28 at 3 11 56 PM
> 
> It shouldn't have border
> 
> image
> image
> 

Sorry, I meant the border on the right, half of it has a border, the other half doesn't

**Comment by InnaAtanasova** [comment]
Impressive job with the table! Once Travis is happy you can go ahead and merge it

**Comment by salarenko** [comment]
Looks good. I'm not sure about one visual thing. Shouldn't the checkbox in collapsed table be aligned to the top? The docs say nothing about it, but thats the case @InnaAtanasova had to cover in very similar List with checkbox layout and it looks much better.

This is how it looks now:
image

This is how it looks with checkbox aligned to the top: (`vertical-align`)
image

**Pull request merged**